### PR TITLE
X2Go Dockerfile

### PR DESCRIPTION
**Pull request by mikedep333**
**This PR is not ready to be pulled yet.** I am opening  this PR early in order to ask a design question.

I would like to create a minimal X2Go Dockerfile, that only contains what users need to run user-installed applications in "Single Application" mode (AKA, "Seamless Window" mode).

I would also like to create a Dockerfile for each desktop environment.

May I base the desktop environment Dockerfiles off of the minimal X2Go Dockerfile?
 (It appears that every existing Fedora Dockerfile is directly based off of the base Fedora Dockerfile.)

### CODE REVIEW

**Comment by znmeb** [comment]
It's not clear what the goal is here. Are you trying to build a Fedora-based Docker image that one can use to run *desktop* applications? If so, there are a few attempts at this, but I think requiring a VNC, Spice or RDP client on the user's end is a non-starter. There's something called "NoVNC" that I've seen in a few "desktop" Docker images. It claims to serve up a desktop that can be viewed in a browser.

There are a few images on Docker Hub that use NoVNC - this one seems to be the simplest:

https://github.com/paimpozhil/docker-novnc

NoVNC itself is at http://kanaka.github.io/noVNC/

**Comment by mikedep333** [comment]
Hi @znmeb. That is what I am trying to do. And to run entire desktop environments.

I do not see why requiring an X2Go Client (the regular X2Go Client, PyHoca-GUI, or PyHoca-CLI) is a non-starter. X2Go has numerous users who are OK with installing an X2Go Client to access physical or virtual machines remotely, myself included. How is that any different from accessing a container remotely? Furthermore, X2Go Client for Windows can be used in portable mode.

The advantage of running X2Go Server in a container is that a privileged user can install all the graphical applications and desktop environments in a container without risking accidentally messing up the configuration of the host OS. Containers may not securely contain yet, but they do prevent you from accidentally messing up the host..

As for why you would use X2Go when VNC has the noVNC client available, one of the advantages of X2Go is extremely good performance over a WAN. Another advantage is built-in handling multiple user sessions via SSH. An HTML5 X2Go client would be nice, but that would require a lot of upstream X2Go work, including improving [an HTML5 X server](https://github.com/GothAck/javascript-x-server/commits/master). It would also be less efficient on Linux because you already have the X.org server running, and you would be running an HTML5 X Server on top of it. (The beauty of X2Go is that it enhances the X11 protocol rather than using a completely unrelated protocol like VNC does.)

That said, I will take a look at [docker-novnc](https://github.com/paimpozhil/docker-novnc) for suggestions. Thanks.

-Mike

**Comment by mikedep333** [comment]
After talking with a few people on IRC, I have created both x2go/x2go-mate and x2go/x2go-minimal. I will rename/move/refactor stuff as needed. I didn't know that if 2 docker files contain the 1st X lines in common, that they will share images until X + 1.

Also, this is still not ready to be merged because I need to fix this text rendering bug in x2go-minimal:
http://imgur.com/ut4GXqE

**Comment by mikedep333** [comment]
I believe this is ready to be merged now. I will add more desktop environments via additional PRs.

**Comment by znmeb** [comment]
If I'm running this image in a container hosted by Boot2Docker on Windows, which VNC client do I need?

**Comment by mikedep333** [comment]
@znmeb X2Go does not use the VNC protocol. It uses its own protocol, which primarily consists of the NX3 protocol.

So you need X2Go Client for Windows. Current version:
http://code.x2go.org/releases/binary-win32/x2goclient/releases/4.0.3.2-20150301/

You can also use PyHoca-GUI for Windows, which runs in the taskbar and is optimized for connecting to multiple X2Go Servers at the same time. Current version:
http://code.x2go.org/releases/binary-win32/pyhoca-gui/releases/0.5.0.4-20150125/

I maintain the Windows ports of both pieces of software :)

**Comment by scollier** [comment]
Hey @mikedep333 

I built the x2go-minimal and ran it both the recommended ways in the README.  However, it's not recognizing my local user, or the user: user | newpass.  Here is the message I'm getting:


$ pyhoca-cli --server localhost -N -c GNOME --user <my_local_user> --add-to-known-hosts
pyhoca-cli[29751] NOTICE: preparing requested X2Go session
pyhoca-cli[29751] (PyHocaCLI) NOTICE: preparing requested X2Go session
pyhoca-cli[29751] (x2gosessregistry-pylib) NOTICE: registering X2Go session Pyhoca-Client_Session...
pyhoca-cli[29751] (x2goclient-pylib) NOTICE: initializing X2Go session...
pyhoca-cli[29751] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:22
pyhoca-cli[29751] (PyHocaCLI) NOTICE: unlock SSH key file (/home/<my_local_user>/.ssh/id_rsa)
Password: 
pyhoca-cli[29751] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:22
pyhoca-cli[29751] (x2goclient-pylib) WARN: HOOK_session_startup_failed: session startup for session profile ,,Pyhoca-Client_Session failed.
pyhoca-cli[29751] (PyHocaCLI) NOTICE: give the X2Go session some time to come up...


Thoughts on this?  Also, sorry, but what are the expected results here?  Should I see Gnome launch?  I'm doing all this on a single host, BTW.

**Comment by mikedep333** [comment]
Hi @scollier ,
You need to run:
`pyhoca-cli --server localhost -p 50000 -N --user <username> --add-to-known-hosts`

LXTerminal should then open up in "Seamless Window" mode.

When using system accounts, you can leave out `--user <username>` in order to connect as your own account.

GNOME is purposely not installed in x2go-minimal, nor is any other desktop environment. By leaving out the -c option, x2goserver defaults to a `TERMINAL` session, which on x2go-minimal is lxterminal.

For x2go-mate, you would run:
`pyhoca-cli --server localhost -p 50000 -N -c MATE --user <username> --add-to-known-hosts`

**Comment by scollier** [comment]
Hey @mikedep333, Just checking in here.

**Comment by mikedep333** [comment]
Hi @scollier . I do most of my open source contributions on the weekends. I will look into the pyhoca-cli error message we discussed on IRC this weekend.

**Comment by scollier** [comment]
@mikedep333 ah, understood.

**Comment by scollier** [comment]
To reproduce the issue:

```
sudo docker run -d -p 50000:22 --privileged=true --volume /etc/passwd:/etc/passwd --volume /etc/shadow:/etc/shadow --volume /etc/group:/etc/group --volume /etc/gshadow:/etc/gshadow --volume /etc/sudoers:/etc/sudoers --volume /home:/home scollier/x2go
```

Then try to connect:

```
pyhoca-cli --server localhost -p 50000 -N --user scollier --add-to-known-hosts
```

And I get a hang:

```


    $ pyhoca-cli --server localhost -p 50000 -N --user scollier --add-to-known-hosts
    pyhoca-cli[30214] NOTICE: preparing requested X2Go session
    pyhoca-cli[30214] (PyHocaCLI) NOTICE: preparing requested X2Go session
    pyhoca-cli[30214] (x2gosessregistry-pylib) NOTICE: registering X2Go session Pyhoca-Client_Session...
    pyhoca-cli[30214] (x2goclient-pylib) NOTICE: initializing X2Go session...
    pyhoca-cli[30214] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:50000
    pyhoca-cli[30214] (PyHocaCLI) NOTICE: unlock SSH key file (/home/scollier/.ssh/id_rsa)
    Password:
    pyhoca-cli[30214] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:50000
```

**Comment by mikedep333** [comment]
The problem is that `/usr/lib64/x2go/x2gosqlitewrapper` is not setgid.
This may be an `x2goserver` packaging issue. I will look into it.

**Comment by mikedep333** [comment]
This may be a re-introduction of this bug:
https://github.com/docker/docker/issues/6828
My current theory is that the capability `FSETID` is missing during `docker build` for some reason.

**Comment by mikedep333** [comment]
@scollier On my Fedora 22 64-bit system, the cause of the problem was SELinux interfering with yum setting setgid during `docker build`. Reported as [RHBZ #1210974] (https://bugzilla.redhat.com/show_bug.cgi?id=1210974)

**Comment by mikedep333** [comment]
lsm5 and I discussed this on #fedora-cloud

The larger issue between SELinux & Docker is complicated and we need Dan Walsh or Lukas to comment. 

But for now, I believe the best approach is to tell users to temporarily enable the SELinux boolean `virt_sandbox_use_all_caps` in README.md before running `docker build`. Then tell them to disable it afterwards. I will update README.md for this.

**Comment by scollier** [comment]
@mikedep333 Thanks. I'll have a look a bit later.  At a conference ATM.

**Comment by scollier** [comment]
Still getting a hang:

$ pyhoca-cli --server localhost -p 50000 -N --user scollier --add-to-known-hosts
pyhoca-cli[6121] NOTICE: preparing requested X2Go session
pyhoca-cli[6121] (PyHocaCLI) NOTICE: preparing requested X2Go session
pyhoca-cli[6121] (x2gosessregistry-pylib) NOTICE: registering X2Go session Pyhoca-Client_Session...
pyhoca-cli[6121] (x2goclient-pylib) NOTICE: initializing X2Go session...
pyhoca-cli[6121] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:50000
pyhoca-cli[6121] (PyHocaCLI) NOTICE: unlock SSH key file (/home/scollier/.ssh/id_rsa)
Password: 
pyhoca-cli[6121] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:50000

**Comment by scollier** [comment]
The last time I did a docker build it was using cache.  I rebuilt with a --no-cache and got a new errror:

```
$ pyhoca-cli --server localhost -p 50000 -N --user scollier --add-to-known-hosts
pyhoca-cli[23115] NOTICE: preparing requested X2Go session
pyhoca-cli[23115] (PyHocaCLI) NOTICE: preparing requested X2Go session
pyhoca-cli[23115] (x2gosessregistry-pylib) NOTICE: registering X2Go session Pyhoca-Client_Session...
pyhoca-cli[23115] (x2goclient-pylib) NOTICE: initializing X2Go session...
pyhoca-cli[23115] (x2gocontrolsession-pylib) NOTICE: connecting to [localhost]:50000
pyhoca-cli: error: SSH host key verification for remote host [localhost]:50000 failed
```

```
$ docker exec 9ad2e05a1bca ls -l /usr/lib64/x2go/x2gosqlitewrapper
-rwxr-sr-x. 1 root brlapi 7296 Feb 24 16:48 /usr/lib64/x2go/x2gosqlitewrapper
```

**Comment by mikedep333** [comment]
Hi @scollier . The SSH Server's host key has been regenerated, per the logic in the Dockerfle. You can fix this with:
`ssh-keygen -R [localhost]:50000`

**Comment by scollier** [comment]
Hey @mikedep333 I was still having issues here.  I'll try to take another look in the next day or so, but the ssh-keygen didn't work.  I'll gather errors and add in a bit.

**Comment by mikedep333** [comment]
Hi scollier,

When testing, please do the following:
1. Run  pyhoca-cli with `--debug --libdebug`. The latter is actually far more important.
2. `ls -l /usr/lib64/x2go/x2gosqlitewrapper`
3. `ls -la /var/lib/x2go`

**Comment by scollier** [comment]
@dustymabe @jasonbrooks One of y'all mind taking a stab at this?  Need to work with @mikedep333 to get this merged.  Thank you.

**Comment by dustymabe** [comment]
@scollier I can try to look at this on the weekend. Will post back here.

**Comment by dustymabe** [comment]
@mikedep333 I haven't forgotten about this. Apologies for the delay.

**Comment by dustymabe** [comment]
@mikedep333 finally getting started on this one. One thing I notice is that you use FROM fedora. We need to use a specific version (tag) for this so that users don't get the new stuff until it has been verified to work. For example I just ran a build and it pulled the Fedora 22 docker base image (now the latest). Can you update to specify the 22 tag and make sure it still works?

**Comment by mikedep333** [comment]
@dustymabe Changed to specify fedora:22. I am still testing though. (For some reason I thought it was pulling F22 instead of F21 all along. Otherwise, I would have tested before pushing.)

**Comment by dustymabe** [comment]
Thanks.. So with my testing the x2go-minimal works fine. x2go-mate doesn't though:

```
Step 12 : RUN yum -y groups install fedora-release-nonproduct fonts input-methods mate-desktop && yum clean all
 ---> Running in 12a80a900d50
Yum command has been deprecated, redirecting to '/usr/bin/dnf -y groups install fedora-release-nonproduct fonts input-methods mate-desktop'.
See 'man dnf' and 'man yum2dnf' for more information.
To transfer transaction metadata from yum to DNF, run:
'dnf install python-dnf-plugins-extras-migrate && dnf-2 migrate'

Last metadata expiration check performed 0:00:05 ago on Sun Jun 14 05:10:56 2015.
Warning: Group 'fedora-release-nonproduct' does not exist.
Dependencies resolved.
Complete!
Yum command has been deprecated, redirecting to '/usr/bin/dnf clean all'.
See 'man dnf' and 'man yum2dnf' for more information.
To transfer transaction metadata from yum to DNF, run:
'dnf install python-dnf-plugins-extras-migrate && dnf-2 migrate'

Cleaning repos: fedora updates
Cleaning up Everything
 ---> 3a751e5a602d
Removing intermediate container 12a80a900d50
Step 13 : RUN yum -y remove anaconda-core && yum clean all
 ---> Running in b41ec71d4eae
Yum command has been deprecated, redirecting to '/usr/bin/dnf -y remove anaconda-core'.
See 'man dnf' and 'man yum2dnf' for more information.
To transfer transaction metadata from yum to DNF, run:
'dnf install python-dnf-plugins-extras-migrate && dnf-2 migrate'

No match for argument: anaconda-core
Error: No packages marked for removal.
INFO[0090] The command [/bin/sh -c yum -y remove anaconda-core && yum clean all] returned a non-zero code: 1
```

**Comment by mikedep333** [comment]
Hi @dustymabe,

I have updated and tested both Dockerfiles with F22

I do not know why this step was failing for you:
```
Step 13 : RUN yum -y remove anaconda-core && yum clean all
```
But the replacement step succeeds for me:
```
Step 14 : RUN dnf -y remove anaconda-core && dnf clean all
```
It results in 86 packages being removed, and that is intentional.

-Mike

**Comment by mikedep333** [comment]
Also,  I and some testers via IRC have run into this bug with pyhoca-cli. It is a harmless error message though:
http://bugs.x2go.org/cgi-bin/bugreport.cgi?bug=877

The regular [x2goclient](http://wiki.x2go.org/doku.php/download:start#the_client-side_of_x2go), which we on the X2Go project recommend for average users, is not affected.

**Comment by dustymabe** [comment]
@mikedep333 Thanks! I'll run through this soon and get back to you. Hopefully I don't hit the same issue I did before.

**Comment by dustymabe** [comment]
@mikedep333 - Ran through the READMEs. Seems to work for me when running with the default user account. I get this error when running with the *docker host's* accounts though:

```
Connection failed DBD::SQLite::db prepare failed: no such table: sessions at /usr/lib64/x2go/x2gosqlitewrapper.pl line 476. Can't call method "execute" on an undefined value at /usr/lib64/x2go/x2gosqlitewrapper.pl line 484.
```

That is output by the x2goclient GUI.

**Comment by mikedep333** [comment]
@dustymabe , I was unable to produce the issue when using the docker host's accounts with the 2 X2Go Fedora-Dockerfiles. I tried both x2goclient and pyhoca-cli on a remote F22 client.

Could you run these 2 commands?
```
docker run <image-name> ls -l /usr/lib64/x2go/x2gosqlitewrapper
docker run <image-name> ls -la /var/lib/x2go
```
(x2gosqlitewrapper should not be confused with x2gosqlitewrapper.pl. The former needs to be setgid, the later should not.)

Also,

I was able to reproduce your error message when I replaced the sqlite db /var/lib/x2go/x2go_sessions with an empty file.

Of course, that should never be an empty file. The spec file's %post section calls the 1-line command to create the db `x2godbadmin --createdb`. 
http://pkgs.fedoraproject.org/cgit/x2goserver.git/tree/x2goserver.spec#n212
(If the db ever gets messed up for some reason, the sysadmin can tell users to disconnect and then re-create it with that 1 command. But for a freshly created docker image, that should not be the base.)

**Comment by dustymabe** [comment]
Here is the output of the commands you requested:

```
[root@kushal x2go-minimal]# docker run dmabe/x2go-minimal ls -l /usr/lib64/x2go/x2gosqlitewrapper
-rwxr-sr-x. 1 root x2gouser 7296 Apr 27 17:38 /usr/lib64/x2go/x2gosqlitewrapper
[root@kushal x2go-minimal]# docker run dmabe/x2go-minimal ls -l /usr/lib64/x2go/x2gosqlitewrapper.pl
-rwxr-xr-x. 1 root root 17820 Apr 27 17:38 /usr/lib64/x2go/x2gosqlitewrapper.pl
[root@kushal x2go-minimal]# docker run dmabe/x2go-minimal ls -la /var/lib/x2go
total 24
drwxrwx---.  2 root x2gouser  4096 Jul 28 03:15 .
drwxr-xr-x. 14 root root      4096 Jul 28 03:25 ..
-rw-rw----.  1 root x2gouser 13312 Jul 28 03:15 x2go_sessions
```

The `/var/lib/x2go/x2go_sessions` file is definitely not empty:

```
[root@e0db2ec22948 /]# echo '.dump' | sqlite3 /var/lib/x2go/x2go_sessions | wc -l 
45
```

**Comment by mikedep333** [comment]
FYI, I was talking with another X2Go developer (ionic) and he is not sure why the DBD::SQLite::db prepare error is being produced. It is a vague error message and I will continue to try to narrow it down.

**Comment by mikedep333** [comment]
@dustymabe  I have decided to delay, or close, this PR.

I will close it in 1 week if I do not receive feedback. I will not delete my branch though.

I've been doing a lot of [self-reflection about myself and the X2Go project](http://mikedep333.blogspot.com/2015/08/confessions-of-open-source-developer.html), and I had conversations at Flock and on IRC with people like @rhatdan & @twaugh .

Here is my reasoning:

1. A single X2Go container accessible by multiple users, with an sqlite database accessible by multiple users (via setgid executables), is orthogonal to the Docker instance vs image model. A container instance should be accessible by only 1 user, and the multiple instances should share that 1 container image. In other words, `Container != VM`.
2. I should add a "single-user mode" or something in upstream X2Go (where setgid is not required), rather than implementing workarounds in this downstream project.
3. This code will be a lot cleaner (and lower maintenance) once SELinux integration with containers is improved, and once [Layered images are available.](https://fedoraproject.org/wiki/Changes/Layered_Docker_Image_Build_Service)

I welcome feedback.

PS: Thanks to @sgallagher for [advice](https://amphibian.com/136) also.

**Comment by mikedep333** [comment]
In other words, I feel like I have been trying to hammer a square peg through a round hole.

This is bad because:
1. I had to spend time hammering. (The issues/testing of this PR.)
2. I have to spend time justifying hammering a square peg through a round hole (my Flock convo with Dan Walsh about multiple users), even after I have already completed doing so.
3. The peg will not be the same as it originally was. For example, I doubt that X2Go printer sharing still works.
4. I have to waste time explaining the last point to users and supporting them when they are not aware of it.
5. I have to spend more time on maintenance in general.

**Comment by znmeb** [comment]
@mikedep333 If it's any consolation, I've greatly scaled back my own ambitions for using Docker. Docker's _raison d'être_ is in the world of servers and DevOps. It solves problems there. 

I'm trying to solve a different problem - portability of robust open-source software to desktop ***end users***, especially Windows.  There isn't an end-user-friendly Docker environment for a Windows user yet.

**Comment by mikedep333** [comment]
Hi @znmeb,

Thanks for the perspective.

Yeah, using a tool for something it wasn't meant for is often part of The Open Source Way. For example, people thought Knoppix was crazy, until lots of other distros started running Live too.

However, The Open Source Way is to fix the integration issues upstream, rather than hacking workarounds downstream.

BartPE was the equivalent of Knoppix. It took the hackaround-downstream approach with Windows XP. And that is why BartPE never worked with Vista or Windows 7; Windows has no upstream open source project for it to integrate with.

If you lack the resources to do either,  you should do neither.

In my case, I do have the resources (free time) to add a single-user-mode to X2Go. But while I do so, I need to to wait for layered Docker Images and better SELinux integration to be completed. Others are already working on them, and are making good progress.

-Mike

File: x2go/x2go-mate/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+RUN sed --in-place=.orig 's/^account    required     pam_nologin.so/#account    required     pam_nologin.so/' /etc/pam.d/sshd
+
+# Install MATE desktop environment
+# Note that these groups are a subset of the groups that group MATE depends on
+#
+# Note: This installs many things we need like systemd as well as things we do
+# not need like the Linux kernel and anaconda.
+RUN dnf -y install fedora-release && dnf clean all
+RUN dnf -y groups install fonts input-methods mate-desktop && dnf clean all
+RUN dnf -y remove anaconda-core && dnf clean all
```
Comment (vpavlin) [inline x2go/x2go-mate/Dockerfile]: Merge the `dnf remove` with the `install` line - if those two are separate basically does not have any effect as the data is still exist in the previous layer even if removed in the next layer.

Comment (mikedep333) [inline x2go/x2go-mate/Dockerfile]: @vpavlin Thanks. I will make this change when I continue to develop these DockerFiles. Right now, I am planning on putting them on hold.

Comment (mikedep333) [inline x2go/x2go-mate/Dockerfile]: @vpavlin , actually. I think the mate-desktop pulls in Anaconda. I need to understand the `group install` logic better.
Either way, I am closing this now.

**Comment by mikedep333** [comment]
As mentioned, I am closing this.

Once all 3 things occur, I will look into reviving it:

1. I am an upstream X2Go developer, so adding a "single-user mode" (working title) is on my to-do list.
2. [Layered images for Fedora-Dockerfiles](https://fedoraproject.org/wiki/Changes/Layered_Docker_Image_Build_Service)
3. Improved integration between SELinux & `docker build` (or the latter being replaced entirely)


But whoever you are, feel free to revive it yourself!

> death isn't the end. [-Melenis (Age of Wonders III)](https://youtu.be/o1RGqYFqTro?t=142)

**Pull request closed without merging**